File: hse/course_03/ProgLangs/01/python/pywidgets.py
from _pywidgets import *

_pointers = {}


_pushButton_onClicked_handlers = {} # PushButton -> receivers

def _pushButton_onClicked_emitter(ptr):
    button = _pointers[ptr]
    for handler in _pushButton_onClicked_handlers[button]:
        handler(button)

PushButton_SetOnClickedStaticHandler(_pushButton_onClicked_emitter)

class Object:
    @property
    def _pointer(self):
        return self.__pointer

    @_pointer.setter
    def _pointer(self, value):
        _pointers[value] = self
        self.__pointer = value

    @_pointer.deleter
    def _pointer(self):
        del _pointers[self.__pointer]
        self.__pointer = None

    _children = []

    def __init__(self):
        raise NotImplementedError("Object is an abstract class")

    def __del__(self):
        if self._pointer:
            Object_Delete(self._pointer)
            del self._pointer

        while len(self._children):
            child = self._children.pop()
            del child._pointer
            del child


    def get_class_name(self):
        return Object_GetClassName(self._pointer)


class Widget(Object):
    def __init__(self, parent=None):
        if parent:
            self._pointer = Widget_New(parent._pointer)
            parent._children.append(self)
        else:
            self._pointer = Widget_New()

    def set_visible(self, v):
        Widget_SetVisible(self._pointer, v)

    def set_window_title(self, title):
        Widget_SetWindowTitle(self._pointer, title)

    def set_layout(self, layout):
        Widget_SetLayout(self._pointer, layout._pointer)

    def set_size(self, w, h):
        Widget_SetSize(self._pointer, w, h)

class PushButton(Widget):
    def __init__(self, parent=None):
        self._pointer = PushButton_New(parent._pointer)
        parent._children.append(self)

        PushButton_RegisterOnClicked(self._pointer)

    def __del__(self):
        Widget.__del__(self)
        _pushButton_onClicked_handlers.pop(self, None)

    def set_text(self, text):
        PushButton_SetText(self._pointer, text)

    def set_on_clicked(self, callback):
        _pushButton_onClicked_handlers[self] = _pushButton_onClicked_handlers.get(self, []) + [callback]


class Label(Widget):
    def __init__(self, parent):
        self._pointer = Label_New(parent._pointer)
        parent._children.append(self)

    def set_text(self, text):
        Label_SetText(self._pointer, text)


class Application(Object):
    def __init__(self):
        self._pointer = Application_New()

    def exec(self):
        return Application_Exec(self._pointer)


class Layout(Object):
    def __init__(self, parent):
        self._pointer = Layout_New(parent._pointer)
        parent._children.append(self)

    def add_widget(self, widget):
        Layout_AddWidget(self._pointer, widget._pointer)


class VBoxLayout(Layout):
    def __init__(self, parent):
        self._pointer = VBoxLayout_New(parent._pointer)
        parent._children.append(self)

File: hse/course_03/ProgLangs/01/c/_pywidgets.c
#include <_pywidgets.h>

void *pointerFromTupleLongElement(PyObject *tuple, Py_ssize_t index) {
    PyObject *obj = PyTuple_GetItem(tuple, index);
    if (obj) {
        if (PyLong_Check(obj)) {
            return PyLong_AsVoidPtr(obj);
        }
    }

    return NULL;
}

static PyObject *_pywidgets_Object_GetClassName(PyObject *self, PyObject *args) {
    struct Object *obj = pointerFromTupleLongElement(args, 0);
    if (!obj) {
        return NULL;
    }
    const char *className = Object_GetClassName(obj);
    return PyUnicode_FromString(className);
}

static PyObject *_pywidgets_Object_Delete(PyObject *self, PyObject *args) {
    struct Object *obj = pointerFromTupleLongElement(args, 0);
    if (!obj) {
        return NULL;
    }

    Object_Delete(obj);
    Py_RETURN_NONE;
}

static PyObject *_pywidgets_Application_New(PyObject *self, PyObject *args) {
    struct Application *app = Application_New();
    return PyLong_FromVoidPtr(app);
}

static PyObject *_pywidgets_Application_Exec(PyObject *self, PyObject *args) {
    struct Application *app = pointerFromTupleLongElement(args, 0);
    if (!app) {
        return NULL;
    }
    int status = Application_Exec(app);
    return PyLong_FromLong(status);
}

static PyObject *_pywidgets_VBoxLayout_New(PyObject *self, PyObject *args) {
    struct Object *parent = pointerFromTupleLongElement(args, 0);
    if (!parent) {
        return NULL;
    }
    struct VBoxLayout *vbox = VBoxLayout_New((struct Widget *) parent);
    return PyLong_FromVoidPtr(vbox);
}

static PyObject *_pywidgets_Layout_AddWidget(PyObject *self, PyObject *args) {
    struct Layout *layout = pointerFromTupleLongElement(args, 0);
    struct Widget *widget = pointerFromTupleLongElement(args, 1);
    if (!layout || !widget) {
        return NULL;
    }

    Layout_AddWidget(layout, widget);
    Py_RETURN_NONE;
}

static PyObject *_pywidgets_Widget_New(PyObject *self, PyObject *args) {
    struct Object *parent = PyTuple_Size(args) ? pointerFromTupleLongElement(args, 0) : NULL;

    struct Widget *widget = Widget_New((struct Widget *) parent);
    return PyLong_FromVoidPtr(widget);
}

static PyObject *_pywidgets_Widget_SetVisible(PyObject *self, PyObject *args) {
    struct Widget *widget = pointerFromTupleLongElement(args, 0);
    PyObject *visibilityObject = PyTuple_GetItem(args, 1);
    if (!widget || !visibilityObject || !PyBool_Check(visibilityObject)) {
        return NULL;
    }
    bool visibility = (bool) PyObject_IsTrue(visibilityObject);

    Widget_SetVisible(widget, visibility);

    Py_RETURN_NONE;
}

static PyObject *_pywidgets_Widget_SetWindowTitle(PyObject *self, PyObject *args) {
    struct Widget *widget = pointerFromTupleLongElement(args, 0);
    PyObject *titleObject = PyTuple_GetItem(args, 1);
    if (!widget || !titleObject || !PyUnicode_Check(titleObject)) {
        return NULL;
    }

    const char *title = PyUnicode_AsUTF8(titleObject);
    Widget_SetWindowTitle(widget, title);
    Py_RETURN_NONE;
}

static PyObject *_pywidgets_Widget_SetLayout(PyObject *self, PyObject *args) {
    struct Widget *widget = pointerFromTupleLongElement(args, 0);
    struct Layout *layout = pointerFromTupleLongElement(args, 1);
    if (!widget || !layout) {
        return NULL;
    }

    Widget_SetLayout(widget, layout);
    Py_RETURN_NONE;
}

static PyObject *_pywidgets_Widget_SetSize(PyObject *self, PyObject *args) {
    struct Widget *widget = pointerFromTupleLongElement(args, 0);
    PyObject *widthObject = PyTuple_GetItem(args, 1);
    PyObject *heightObject = PyTuple_GetItem(args, 2);
    if (!widget || !widthObject || !heightObject || !PyLong_Check(widthObject) || !PyLong_Check(heightObject)) {
        return NULL;
    }

    long width = PyLong_AsLong(widthObject);
    long height = PyLong_AsLong(heightObject);

    Widget_SetSize(widget, (int) width, (int) height);
    Py_RETURN_NONE;
}

static PyObject *_pywidgets_Label_New(PyObject *self, PyObject *args) {
    struct Object *parent = pointerFromTupleLongElement(args, 0);
    if (!parent) {
        return NULL;
    }

    struct Label *label = Label_New((struct Widget *) parent);
    return PyLong_FromVoidPtr(label);
}

static PyObject *_pywidgets_Label_SetText(PyObject *self, PyObject *args) {
    struct Label *label = pointerFromTupleLongElement(args, 0);
    PyObject *textObject = PyTuple_GetItem(args, 1);
    if (!label || !textObject || !PyUnicode_Check(textObject)) {
        return NULL;
    }
    const char *text = PyUnicode_AsUTF8(textObject);

    Label_SetText(label, text);

    Py_RETURN_NONE;
}

static PyObject *_pywidgets_PushButton_New(PyObject *self, PyObject *args) {
    struct Widget *parent = pointerFromTupleLongElement(args, 0);
    if (!parent) {
        return NULL;
    }

    struct PushButton *button = PushButton_New(parent);
    return PyLong_FromVoidPtr(button);
}

static PyObject *_pywidgets_PushButton_SetText(PyObject *self, PyObject *args) {
    struct PushButton *button = pointerFromTupleLongElement(args, 0);
    PyObject *textObject = PyTuple_GetItem(args, 1);
    if (!button || !textObject || !PyUnicode_Check(textObject)) {
        return NULL;
    }
    const char *text = PyUnicode_AsUTF8(textObject);

    PushButton_SetText(button, text);

    Py_RETURN_NONE;
}

static PyObject *pushButtonOnClickedHandler;

static PyObject *_pywidgets_PushButton_SetOnClickedStaticHandler(PyObject *self, PyObject *args) {
    PyObject *callObject = PyTuple_GetItem(args, 0);
    if (!callObject || !PyCallable_Check(callObject)) {
        return NULL;
    }
    pushButtonOnClickedHandler = callObject;

    Py_RETURN_NONE;
}

void onClicked(struct Object *obj) {
    PyObject_Call(pushButtonOnClickedHandler, PyTuple_Pack(1, PyLong_FromVoidPtr(obj)), NULL);
}

static PyObject *_pywidgets_PushButton_RegisterOnClicked(PyObject *self, PyObject *args) {
    struct PushButton *button = pointerFromTupleLongElement(args, 0);
    if (!button) {
        return NULL;
    }

    PushButton_SetOnClicked(button, onClicked);

    Py_RETURN_NONE;
}